x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ExtractStatsCommonFilter.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules.logical;
+
+import org.elasticsearch.xpack.esql.core.expression.Alias;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.Literal;
+import org.elasticsearch.xpack.esql.core.expression.NamedExpression;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.AggregateFunction;
+import org.elasticsearch.xpack.esql.plan.logical.Aggregate;
+import org.elasticsearch.xpack.esql.plan.logical.Filter;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.xpack.esql.core.expression.predicate.Predicates.combineAnd;
+import static org.elasticsearch.xpack.esql.core.expression.predicate.Predicates.inCommon;
+import static org.elasticsearch.xpack.esql.core.expression.predicate.Predicates.splitAnd;
+import static org.elasticsearch.xpack.esql.core.expression.predicate.Predicates.subtract;
+
+/**
+ * Extract a per-function expression filter applied to all the aggs as a query {@link Filter}, when no groups are provided.
+ * <p>
+ *     Example:
+ *     <pre>
+ *         ... | STATS MIN(a) WHERE b > 0, MIN(c) WHERE b > 0 | ...
+ *         =>
+ *         ... | WHERE b > 0 | STATS MIN(a), MIN(c) | ...
+ *     </pre>
+ */
+public final class ExtractStatsCommonFilter extends OptimizerRules.OptimizerRule<Aggregate> {
```

Comment:
ExtractAggregateCommonFilter since Stats is the name of the command but inside the Planner we use Aggregate.

Reply:
Sure, will rename.
 
 But just in case we might want to be consistent: 
* previously we had `ReplaceStatsAggExpressionWithEval` (added in #104958), refactored (away) in the meantime, but we still name it a few times in [stats_percentile.csv-spec](https://github.com/elastic/elasticsearch/blob/be43c97ff361eb5c6907fa262e820bdc3ae3c205/x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats_percentile.csv-spec#L80).
* currently we still have `RemoveStatsOverride` and `PruneOrderByBeforeStats`; we _should_ rename those too at some point indeed, since they'll apply to other (inline)stats.